linux_os/guide/services/ssh/ssh_server/sshd_set_maxstartups/oval/shared.xml

```diff
@@ -0,0 +1,30 @@
+<def-group>
+  <definition class="compliance" id="sshd_set_maxstartups" version="1">
+    {{{ oval_metadata("Ensure 'MaxStartups' is configured in '/etc/ssh/sshd_config'") }}}
+    <criteria comment="sshd is configured correctly or is not installed"
+    operator="OR">
+      <criteria comment="sshd is not installed" operator="AND">
+        <extend_definition comment="sshd is not required or requirement is unset"
+        definition_ref="sshd_not_required_or_unset" />
+        {{% if product in ['opensuse', 'sle12', 'sle15'] %}}
+        <extend_definition comment="rpm package openssh removed"
+        definition_ref="package_openssh_removed" />
+        {{% else %}}
+        <extend_definition comment="rpm package openssh-server removed"
+        definition_ref="package_openssh-server_removed" />
+        {{% endif %}}
+      </criteria>
+      <criterion comment="Check the MaxStartups value in /etc/ssh/sshd_config"
+      test_ref="test_sshd_set_maxstartups" />
+    </criteria>
+  </definition>
+  
+  {{{ oval_line_in_file_test(path='/etc/ssh/sshd_config', parameter='MaxStartups') }}}
+  {{{ oval_line_in_file_object(path='/etc/ssh/sshd_config', prefix_regex="^[ \\t]*(?i)", parameter='MaxStartups', separator_regex='(?-i)[ \\t]+') }}}
+  
+  <ind:textfilecontent54_state id="state_sshd_set_maxstartups" version="1">
```

Comment:
Just a general comment... It makes this OVAL check much harder, but technically CIS is worded to allow inequality comparisons... It would require splitting the variable into three parts (using local variables and pattern matching) -- but would allow a stricter-than-necessary config to pass. However, it is hard to determine intent -- if the admin specifically set a value and the existing value was stricter, the rule would pass even though the admin might want to still remediate (thought they could always do this manually by generating a profile-based remediation without scanning).

Thoughts?

Reply:
I agree with the idea that it would be sensible to include a test that allows stricter-than-necessary rules to pass the test.

I do also see what you're suggesting though, that an admin might want to end a remediation run with the CIS-recommended value of `10:30:60` regardless of whether a more strict value is already in place. But reading the wording of the CIS benchmark, CIS does not appear to be too opinionated about the specific value here so I think maybe it's fine for a validation run to consider a stricter value as a pass.

I'll have to look closer into splitting the variable and validating the separate segments to allow the stricter configs to pass the rule. I'll have to look through other tests and see whether I can find any which already do something similar.

Reply:
Well, I don't have the answer to the local variable part, but the rest of it would look something like:


```xml
<def-group>
  <definition class="compliance" id="{{{ rule_id }}}" version="1">
    {{{ oval_metadata("The MaxStartups parameter specifies the maximum number of concurrent unauthenticated connections to the SSH daemon.") }}}
    <criteria operator="OR">
      <extend_definition definition_ref="package_openssh-server_removed" negate="true" comment="SSH daemon is not installed" applicability_check="true" />
      <criteria operator="AND">
        <criterion test_ref="tst_maxstartups_start_parameter" comment="SSH MaxStartups start parameter is less than or equal to 10" />
        <criterion test_ref="tst_maxstartups_rate_parameter" comment="SSH MaxStartups rate parameter is greater than or equal to 30" />
        <criterion test_ref="tst_maxstartups_full_parameter" comment="SSH MaxStartups full parameter is less than or equal to 100" />
      </criteria>
    </criteria>
  </definition>
  <ind:textfilecontent54_object id="obj_sshd_config_maxstartups_first_parameter" version="1">
    <ind:filepath operation="equals">/etc/ssh/sshd_config</ind:filepath>
    <ind:pattern operation="pattern match" datatype="string">(?i)^\s*MaxStartups\s+(\d+):\d+:\d+\s*$</ind:pattern>
    <ind:instance operation="greater than or equal" datatype="int">1</ind:instance>
  </ind:textfilecontent54_object>
  <ind:textfilecontent54_object id="obj_sshd_config_maxstartups_second_parameter" version="1">
    <ind:filepath operation="equals">/etc/ssh/sshd_config</ind:filepath>
    <ind:pattern operation="pattern match" datatype="string">(?i)^\s*MaxStartups\s+\d+:(\d+):\d+\s*$</ind:pattern>
    <ind:instance operation="greater than or equal" datatype="int">1</ind:instance>
  </ind:textfilecontent54_object>
  <ind:textfilecontent54_object id="obj_sshd_config_maxstartups_third_parameter" version="1">
    <ind:filepath operation="equals">/etc/ssh/sshd_config</ind:filepath>
    <ind:pattern operation="pattern match" datatype="string">(?i)^\s*MaxStartups\s+\d+:\d+:(\d+)\s*$</ind:pattern>
    <ind:instance operation="greater than or equal" datatype="int">1</ind:instance>
  </ind:textfilecontent54_object>
  <ind:textfilecontent54_state id="ste_sshd_config_start_parameter_valid" version="1">
    <ind:subexpression datatype="int" operation="less than or equal">10</ind:subexpression>
  </ind:textfilecontent54_state>
  <ind:textfilecontent54_state id="ste_sshd_config_rate_parameter_valid" version="1">
    <ind:subexpression datatype="int" operation="greater than or equal">30</ind:subexpression>
  </ind:textfilecontent54_state>
  <ind:textfilecontent54_state id="ste_sshd_config_full_parameter_valid" version="1">
    <ind:subexpression datatype="int" operation="less than or equal">100</ind:subexpression>
  </ind:textfilecontent54_state>
  <ind:textfilecontent54_test check="all" check_existence="only_one_exists" id="tst_maxstartups_start_parameter" version="1" comment="SSH MaxStartups start parameter is less than or equal to 10">
    <ind:object object_ref="obj_sshd_config_maxstartups_first_parameter" />
    <ind:state state_ref="ste_sshd_config_start_parameter_valid" />
  </ind:textfilecontent54_test>
  <ind:textfilecontent54_test check="all" check_existence="only_one_exists" id="tst_maxstartups_rate_parameter" version="1" comment="SSH MaxStartups rate parameter is greater than or equal to 30">
    <ind:object object_ref="obj_sshd_config_maxstartups_second_parameter" />
    <ind:state state_ref="ste_sshd_config_rate_parameter_valid" />
  </ind:textfilecontent54_test>
  <ind:textfilecontent54_test check="all" check_existence="only_one_exists" id="tst_maxstartups_full_parameter" version="1" comment="SSH MaxStartups full parameter is less than or equal to 100">
    <ind:object object_ref="obj_sshd_config_maxstartups_third_parameter" />
    <ind:state state_ref="ste_sshd_config_full_parameter_valid" />
  </ind:textfilecontent54_test>
</def-group>
```

I think you're looking for [something like this](https://github.com/ComplianceAsCode/content/blob/master/linux_os/guide/system/accounts/accounts-restrictions/root_logins/no_shelllogin_for_systemaccounts/oval/shared.xml#L143-L150) for the variable parts. :-) 

(Above OVAL shared with permission from internal contributors, feel free to use all or part of it here in this CaC PR). 

If it is more than you want to bite off, just LMK and I can open a second PR after this merges.

Reply:
I see #7073 has been merged now so technically this PR could be as well I think, but I still plan on tackling this one when I get a chance. I like a challenge...

Reply:
@cipherboy I've tried to integrate this now. Hopefully I understood the local variable part properly. The regex there should match the value assigned to the last instance of `MaxStartups` found in the `/etc/ssh/sshd_config` file.

Your test lines above seemed to make sense as-is, so I've incorporated that mostly verbatim into the OVAL file.